### ISSUE TITLE
Karaoke generation support

### ISSUE THREAD

**Issue by ahuei123456** [xIceArcher/lyricsheets#1]
As previously discussed, I propose the addition of support for marking lines that will have Aegisub's karaoke templater applied to them, which would include the following features:

1. Using column 'D' in the google sheets to flag karaoke lines—Each separate template will have a different flag, to be formatted as K{0/1...}.
![image](https://user-images.githubusercontent.com/984601/176984492-49f4c2a7-040f-44ea-8353-aed07c590e8e.png)
2. Generating a new style for karaoke lines—I propose to name it "Song Title - K{0/1/...)"
3. Automatically commenting karaoke lines
![image](https://user-images.githubusercontent.com/984601/176984486-e5f68a5e-843b-4703-ab4b-2ce09b9678de.png)
4. Generate inline-fx for actor switches during a line and placing it within the \kf tag
![image](https://user-images.githubusercontent.com/984601/176984690-309c766c-b073-402f-afa3-95356959b2ff.png)
5. Adding a new modifier flag to specify when karaoke templater needs to be run on a song

Potential improvements

- For 2), maybe there could be a better name for the style.
- For 4), putting character names as inline-fx instead of just numbers would be more descriptive (for the karaoke template). The names could possibly be placed in row 2 of the gsheets, under the color specifications
![image](https://user-images.githubusercontent.com/984601/176984851-f4135630-425c-4db9-be42-86d468230824.png)
- For 5), no matter how many songs are in an .ass that have the modifier flag, when it is being populated, the karaoke templater should only be run _once_, for performance reasons.

**Comment by xIceArcher** [comment]
> 1. Using column 'D' in the google sheets to flag karaoke lines—Each separate template will have a different flag, to be formatted as K{0/1...}.

LGTM, can move it to another column later (I'm thinking to swap columns E and F afterwards, but that can be done after the issue is closed)

> 2. Generating a new style for karaoke lines—I propose to name it "Song Title - K{0/1/...)"

Currently when populating songs, in the pre-processing step, all the lines with a style with prefix `Song` (excluding lines with the style `Song` itself) are removed and regenerated (https://github.com/xIceArcher/lyricsheets/blob/master/populate_songs.py#L76).

Since presumably these lines would need to be regenerated as well (in cases where, e.g. the romaji changed/the karaoke timing of some syllables changed), I suggest adding "Song - " as a prefix as well.

> 3. Automatically commenting karaoke lines

LGTM

> 4. Generate inline-fx for actor switches during a line and placing it within the \kf tag

If this is the convention typically used for Aegisub lua scripts, sure, but maybe it could have a more descriptive tag name than `/-`? Maybe `/actor` if possible. Minor nitpick though.

> Putting character names as inline-fx instead of just numbers would be more descriptive 

Doable too, can even consider just replacing the actor numbers with more descriptive strings if you want, though it might look a bit crowded on the Google sheets (which would still be an issue if row 2 was used).

> 5. Adding a new modifier flag to specify when karaoke templater needs to be run on a song

Should be doable, i.e., run the karaoke templater if any one song in the file requires it.

**Comment by ahuei123456** [comment]
> Since presumably these lines would need to be regenerated as well (in cases where, e.g. the romaji changed/the karaoke timing of some syllables changed), I suggest adding "Song - " as a prefix as well.

How about `Song - {Song Title} {K0/1...}`?

> If this is the convention typically used for Aegisub lua scripts, sure, but maybe it could have a more descriptive tag name than /-? Maybe /actor if possible.

Honestly a good point, but that's the format that Aegisub's [karaoke templater](https://aegi.vmoe.info/docs/3.0/Karaoke_inline-fx/) uses so unfortunately we're stuck with it.

> though it might look a bit crowded on the Google sheets (which would still be an issue if row 2 was used)

What about putting the names on row 2 of the hidden columns?

**Comment by xIceArcher** [comment]
> > Since presumably these lines would need to be regenerated as well (in cases where, e.g. the romaji changed/the karaoke timing of some syllables changed), I suggest adding "Song - " as a prefix as well.
> 
> How about `Song - {Song Title} {K0/1...}`?
> 
> > If this is the convention typically used for Aegisub lua scripts, sure, but maybe it could have a more descriptive tag name than /-? Maybe /actor if possible.
> 
> Honestly a good point, but that's the format that Aegisub's [karaoke templater](https://aegi.vmoe.info/docs/3.0/Karaoke_inline-fx/) uses so unfortunately we're stuck with it.
> 
> > though it might look a bit crowded on the Google sheets (which would still be an issue if row 2 was used)
> 
> What about putting the names on row 2 of the hidden columns?

All seems OK to me!

**Comment by ahuei123456** [comment]
For #5, the structure of the current program doesn't seem to lend itself well to propagating a modifier flag up. If `populate_songs.py` is called, it directly acts on the styles and events of the input .ass, so if a modifier is found within `modify_song()`, the only data returned to `main()` is the list of events after all the modifiers are executed onto the song. Any ideas on how to propagate the modifier all the way up to `main()` (since aegisub-cli can realistically only be called after the final .ass is written)?

**Comment by xIceArcher** [comment]
> Any ideas on how to propagate the modifier all the way up to main()

Hmm, probably not the most elegant of solutions but could make populate_songs() return multiple values? So for every song, if the modifier list of this song contains this modifier return `outEvents, True` or something similar.

Agreed that it's not a very pretty solution though, since the original assumption was that for each song, the modifiers will (and should) only apply to the scope of the song's events.

**Comment by xIceArcher** [comment]
Thinking about it a little more, am I right to assume that if the karaoke templater is run on a file where not a single line has the effect `karaoke`, then the file should remain unchanged?

If that is the case, I think we could make it slightly simpler by just running the karaoke templater every single time (regardless of whether or not any song requires it). If this is possible, then a song's list of modifiers can continue to be kept within the scope of a song's events (i.e. whatever additional annotations that are needed by the karaoke templater in each song).

I'm assuming running the karaoke templater once shouldn't have much performance impact. Not like performance is a critical aspect of this project anyway.

**Comment by ahuei123456** [comment]
Yeah, if the karaoke templater has no lines to act on then it won't change. Not sure how it'll work in `aegisub-cli` but for regular aegisub it just stays the same, so that sounds good.
